### DISCUSSION TITLE
Advice on Bottom-Up configuration vs. further labeling for domain shifts in multi-animal tracking (Resident-Intruder)

### DISCUSSION THREAD

**Discussion by saareu** [talmolab/sleap#2721] [Help!] [answered]
Hello SLEAP community,

I am seeking advice on optimizing my training and inference configurations versus expanding my dataset.

As part of developing a computational behavioral analysis pipeline for resident-intruder assays, I am tracking two mice: one smaller/younger, and one larger/older that is implanted with a headstage. The skeleton consists of 5 nodes: nose, neck, spine_base, tail_base, and an additional headstage_led for the implanted mouse.

Currently, I have a robust dataset of ~2,200 labeled frames across 7 videos. Videos are preprocessed (padded to maintain constant height/width, zoomed/rotated to display only the cage, re-encoded to mp4v at ~120 fps).

Due to the heavy overlapping of the mice during social interactions, I moved from a top-down approach to a bottom-up model, which generally handles the overlap better. Inference works well on videos that resemble the training settings. However, when running inference on new experiments with slight domain shifts (specifically, brighter lighting and a modified headstage structure), the performance degrades significantly.

Observed Issues in New Experiments:

Nodes are localized incorrectly.

A single mouse's skeleton splits into two separate skeletons.

Partial distinct skeletons merge into one.

Multiple different skeletons overlap onto the same mouse.

Overall tracking identity preservation is unsatisfactory.

(I have attached images demonstrating these failure modes).

My Dilemma:

Given that I already have 2,000+ labeled frames representing a solid baseline of scenes, and the inter-experiment settings don't change drastically, I am wondering if this is a hyperparameter/augmentation issue or strictly a data diversity issue.

Should I focus my efforts on optimizing the bottom-up training and tracker configurations (attached below), or is the standard protocol here to jump straight into adding more labels from these new, brighter/modified videos?

Current Config Highlights:

Backbone: UNet (64 filters, 1 stack)

Augmentations: Intensity and geometric enabled (brightness 0.8–1.2, full rotation).

Tracker: Hungarian matching, window size 24, max instances 4.

(Full pose.yaml and bottomup.yaml attached).

Any insights on adjusting the PAF/Confmap parameters, tracker settings, or augmentation ranges to better generalize across these variations would be highly appreciated!
[bottomup.yaml](https://github.com/user-attachments/files/28308968/bottomup.yaml)
[pose.yaml](https://github.com/user-attachments/files/28308970/pose.yaml)

<img width="709" height="270" alt="image" src="https://github.com/user-attachments/assets/c1d88a1f-75a1-4282-a0af-c2858ca6c72e" />
<img width="865" height="982" alt="image" src="https://github.com/user-attachments/assets/31a14c2a-66b9-4ae1-aeae-7a00f0a5ca78" />

<img width="609" height="525" alt="image" src="https://github.com/user-attachments/assets/35decfd2-943b-41b5-a984-c90e455cbfec" />

<img width="655" height="325" alt="image" src="https://github.com/user-attachments/assets/4e0f7e4d-de98-48b3-a3bb-183c86e3b8ec" />

<img width="517" height="430" alt="image" src="https://github.com/user-attachments/assets/87d773ee-4232-49b9-9caf-43424cc7e713" />

<img width="702" height="467" alt="image" src="https://github.com/user-attachments/assets/f41249a5-a976-48de-907a-6f465014ff31" />

**Comment by tom21100227** [comment, marked as answer]
Hi @saareu , 

> Should I focus my efforts on optimizing the bottom-up training and tracker configurations (attached below), or is the standard protocol here to jump straight into adding more labels from these new, brighter/modified videos?

The answer is **a mixture of both**, to ensure the best tracking quality, adding some labels from the newer video would be best. However there is also some easy improvements that can be done with the existing model. This is indeed a domain-shift problem. 

To make your model more robust to changing lighting and camera perspective condition I would change your augmentation to be more aggressive, the numbers below are just some example numbers, you should double check with [the API reference](https://nn.sleap.ai/latest/configuration/data/#intensityconfig-augmentation_configintensity) and [the interactive tool (only some augmentations are supported in this)](https://nn.sleap.ai/latest/configuration/config-picker/) to determine what's best to simulate your new lightning condition with older data. 

**It is worth noting that all this augmentation could only help with brightness/perspective of new videos**, the model still needs new labels on new videos to perform better on mice with those modified headstage structure. 

```yaml
...
  augmentation_config:
    intensity:
      brightness_min: 0.6      # was 0.8  — widen to bracket the brighter new videos
      brightness_max: 1.5      # was 1.2  (max allowed is 2.0)
      brightness_p: 0.9        # was 0.5  — apply it more often
      contrast_min: 0.6        # was 0.7
      contrast_max: 1.4        # was 1.3
      contrast_p: 0.75         # was 0.5
      gaussian_noise_std: 0.02 # was 0.01 (this is also the repo default)
      gaussian_noise_p: 0.75   # was 0.5
    geometric:
      scale_min: 0.8           # was 0.9  — more size variation
      scale_max: 1.2           # was 1.1
      translate_width: 0.05    # was 0.0  — add positional jitter
      translate_height: 0.05   # was 0.0
      rotation_min: -180.0     # unchanged — already full rotation
      rotation_max: 180.0      # unchanged
```

**As for split/merged/extra skeletons**, in `pose.yaml` I am seeing `max_instance: 4` and `max_tracks: 4`. If there's only ever going to be 2 mice in the cage consider switching to `max_instance: 2` and `max_tracks: 3` (leave room to prevent tracks switching). 

Hope this helps! 

Best,
Tom
